### PR TITLE
Fix the race condition by confirming creation/deletion of machine objects

### LINKED ISSUE #245: Find better solution for solving race conditions with MachineSet controller reconciling
The current MachineSet controller watches machine and determines whether to reconcile the machine set that owns the machine.
This causes race conditions when the MachineSet creates machines, which may trigger another reconciliation, which ends up creating or deleting additional machines as the system may not have detected the modifications from the initial reconciliation.
A mutex lock was added initially to solve the issue, but back-to-back reconciliation happens quick enough such that the race condition still occurs.
A temp fix to sleep is currently proposed to avoid prevent this race condition.

The replica set controller solves this with using expectations, https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/controller_utils.go#L113
This may be something we will want to adopt as a better fix.

### PR DESCRIPTION
**Pull request by k4leung4**
By waiting for the machine objects to be either created or deleted
before leaving Reconcile loop will prevent the race condition of stale
cache not reporting correctly the change to machine objects.

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #245

**Special notes for your reviewer**:

**Release note**:

```release-note

```

@kubernetes/kube-deploy-reviewers

### CODE REVIEW

File: pkg/controller/machineset/controller.go [hunk 1]
```diff
@@ -34,14 +35,18 @@ import (
...
 
-// reconcileMutexSleepSec is the duration to sleep before releasing the mutex lock that is held for reconcilation.
-// See https://github.com/kubernetes-sigs/cluster-api/issues/245
-var reconcileMutexSleepSec = time.Second
+// stateConfirmationTimeout is the amount of time allowed to wait for desired state.
+var stateConfirmationTimeout = 10 * time.Second
+
+// stateConfirmationInterval is the amount of time between polling for the desired state.
+// The polling is against a local memory cache.
+var stateConfirmationInterval = 100 * time.Millisecond
```
Comment (karan) [inline pkg/controller/machineset/controller.go]: Is this too fast?

File: pkg/controller/machineset/controller.go [hunk 2]
```diff
@@ -342,3 +332,41 @@ func getMachinesToDelete(filteredMachines []*v1alpha1.Machine, diff int) []*v1al
...
+		pollErr := util.Poll(stateConfirmationInterval, stateConfirmationTimeout, func() (bool, error) {
+			_, err := c.machineLister.Machines(machine.Namespace).Get(machine.Name)
+			glog.Error(err)
+			if err == nil {
+				return true, nil
+			}
+			if errors.IsNotFound(err) {
+				return false, nil
+			}
+			return false, err
```
Comment (karan) [inline pkg/controller/machineset/controller.go]: Please log all errors explicitly for debugging

File: pkg/controller/machineset/controller.go (hunk 1 above)
Comment (k4leung4) [reply pkg/controller/machineset/controller.go]: this is against an local internal cache, it shouldnt be an issue. 
i can raise the interval if it is a concern

File: pkg/controller/machineset/controller.go (hunk 2 above)
Comment (k4leung4) [reply pkg/controller/machineset/controller.go]: error from the retry loop goes to pollErr which is always logged.

**Comment by karan** [comment]
lgtm

**Comment by medinatiger** [comment]
lgtm

File: pkg/controller/machineset/controller.go [hunk 3]
```diff
@@ -342,3 +332,41 @@ func getMachinesToDelete(filteredMachines []*v1alpha1.Machine, diff int) []*v1al
 	// see: https://github.com/kubernetes/kube-deploy/issues/625
 	return filteredMachines[:diff]
 }
+
+func (c *MachineSetControllerImpl) waitForMachineCreation(machineList []*v1alpha1.Machine) error {
```
[the remark this answers is not in the document]
Comment (k4leung4) [reply pkg/controller/machineset/controller.go]: agree.
will clean up in future PR.

**Pull request merged**